### PR TITLE
[CLOUDTRUST-3862] Roles management

### PR DESCRIPTION
**Pull request by basbeu**

### CODE REVIEW

File: pkg/management/component.go [hunk 1]
```diff
@@ -827,7 +841,25 @@ func (c *component) getUserRolesAsMap(ctx context.Context, realmName, userID str
...
+	userRoles, err := c.getRolesWithAttributes(accessToken, realmName, roles)
+	if err != nil {
+		c.logger.Info(ctx, "msg", "Failed to get user roles", "realm", realmName, "user")
+		return nil, err
+	}
+
+	return c.rolesToMap(userRoles), nil
+}
+
+func (c *component) getRolesWithAttributes(accessToken string, realmName string, roles []kc.RoleRepresentation) ([]kc.RoleRepresentation, error) {
```
Comment (basbeu) [inline pkg/management/component.go:853]: This way of retrieving Roles attributes is only used when we retrieve role mapping on a user. Keycloak does not offer to retrieve role-mapping directly with attributes. We may consider to extend the keycloak api to offer to retrieve in one call the role-mappings and their attributes. The question is do we will have user having a lot of roles ? Is there a need of scalability ?

File: pkg/management/component.go [hunk 2]
```diff
@@ -1351,20 +1383,30 @@ func (c *component) GetRoles(ctx context.Context, realmName string) ([]api.RoleR
...
-		var roleRep api.RoleRepresentation
-		roleRep.ID = roleKc.ID
-		roleRep.Name = roleKc.Name
-		roleRep.Composite = roleKc.Composite
-		roleRep.ClientRole = roleKc.ClientRole
-		roleRep.ContainerID = roleKc.ContainerID
-		roleRep.Description = roleKc.Description
-
-		rolesRep = append(rolesRep, roleRep)
+		if c.isBusinessRole(roleKc) {
```
Comment (fperot74) [inline pkg/management/component.go:1386]: Factorize? You can write convert functions in api/management

File: pkg/management/component.go [hunk 3]
```diff
@@ -785,18 +790,25 @@ func (c *component) GetRolesOfUser(ctx context.Context, realmName, userID string
...
-		roleRep.ID = roleKc.ID
-		roleRep.Name = roleKc.Name
-		roleRep.Composite = roleKc.Composite
-		roleRep.ClientRole = roleKc.ClientRole
-		roleRep.ContainerID = roleKc.ContainerID
-		roleRep.Description = roleKc.Description
-
-		rolesRep = append(rolesRep, roleRep)
+	for _, roleKc := range userRoles {
+		if c.isBusinessRole(roleKc) {
```
Comment (fperot74) [inline pkg/management/component.go:801]: Factorize? You can write convert functions in api/management

**Review by fperot74 — APPROVED** [review]

File: api/management/swagger-api_management.yaml [hunk 4]
```diff
@@ -1165,6 +1165,51 @@ paths:
...
+        schema:
+          type: string
+      - name: roleID
+        in: path
+        description: role id
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
```
[the remark this answers is not in the document]
Comment (basbeu) [reply api/management/swagger-api_management.yaml:1211]: For the two new endpoints for creating and deleting roles, I did the exact same thing that what was done for the delete and create group endpoints. For sure maybe 204 is better, but for consistence I propose to also update the groups endpoint accordingly.

File: api/management/swagger-api_management.yaml [hunk 5]
```diff
@@ -167,7 +167,7 @@ paths:
           type: boolean
       - name: termsOfUse
         in: query
-        description: tells if a terms of use are required (default: false)
+        description: tells if a terms of use are required (default false)
```
[the remark this answers is not in the document]
Comment (basbeu) [reply api/management/swagger-api_management.yaml:170]: Ok

File: configs/keycloak_bridge.yml [hunk 6]
```diff
@@ -175,7 +175,6 @@ rate-management-status: 2
 rate-statistics: 1000
 rate-events: 1000
 rate-register: 1000
-rate-tasks: 10
```
[the remark this answers is not in the document]
Comment (basbeu) [reply configs/keycloak_bridge.yml:178]: Ok

File: pkg/management/component.go (hunk 2 above)
Comment (basbeu) [reply pkg/management/component.go:1386]: Good catch ! Done !

File: pkg/management/component.go (hunk 3 above)
Comment (basbeu) [reply pkg/management/component.go:801]: Good catch ! Done !

File: api/management/swagger-api_management.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (fperot74) [reply api/management/swagger-api_management.yaml:1211]: Yes for old methods... but most recent methods are using 204 like /realms/{realm}/users/{userID}/roles/{roleID}

File: pkg/management/component.go [hunk 7]
```diff
@@ -1386,6 +1416,50 @@ func (c *component) GetRole(ctx context.Context, realmName string, roleID string
...
+
+	role, err := c.keycloakClient.GetRole(accessToken, realmName, roleID)
+	if err != nil {
+		c.logger.Warn(ctx, "err", err.Error())
+		return err
+	}
+
+	var roleName = *role.Name
+
+	err = c.keycloakClient.DeleteRole(accessToken, realmName, roleID)
```
Comment (harture) [inline pkg/management/component.go:1451]: We must ensure we are deleting a business role

File: api/management/api.go [hunk 8]
```diff
@@ -464,6 +464,33 @@ func ConvertUpdatableToKCUser(user UpdatableUserRepresentation) kc.UserRepresent
...
+	roleRep.Name = role.Name
+	roleRep.Composite = role.Composite
+	roleRep.ClientRole = role.ClientRole
+	roleRep.ContainerID = role.ContainerID
+	roleRep.Description = role.Description
+	return roleRep
+}
+
+func ConvertToKCRole(role RoleRepresentation) kc.RoleRepresentation {
+	attributes := map[string][]string{
```
Comment (harture) [inline api/management/api.go:479]: I think we should not hard code this here. the aims of API is to do a conversion only, here we are adding a flag silently.

File: pkg/management/component.go [hunk 9]
```diff
@@ -1386,6 +1416,50 @@ func (c *component) GetRole(ctx context.Context, realmName string, roleID string
 	return roleRep, nil
 }
 
+func (c *component) CreateRole(ctx context.Context, realmName string, role api.RoleRepresentation) (string, error) {
+	var accessToken = ctx.Value(cs.CtContextAccessToken).(string)
+
+	var roleRep = api.ConvertToKCRole(role)
```
Comment (harture) [inline pkg/management/component.go:1422]: Add Business flag here instead of into conversion method

File: go.mod [hunk 10]
```diff
@@ -21,9 +21,9 @@ require (
 	github.com/cenkalti/backoff v2.2.1+incompatible // indirect
 	github.com/certifi/gocertifi v0.0.0-20210429200110-83314bf6d27c // indirect
 	github.com/clbanning/x2j v0.0.0-20191024224557-825249438eec // indirect
-	github.com/cloudtrust/common-service v1.2.6-0.20211110103624-faf08bf0ed05
+	github.com/cloudtrust/common-service v1.2.6-0.20211125171259-4b28d4420e64
 	github.com/cloudtrust/httpclient v1.0.5-0.20211112142708-1160ef94141a
-	github.com/cloudtrust/keycloak-client v1.3.2-0.20211119145053-743d2b2c08dc
+	github.com/cloudtrust/keycloak-client v1.3.2-0.20211125170427-bb3b9f3a1227
```
Comment (harture) [inline go.mod:26]: How the number can change while having same version number ? sound strange to me...

**Review by harture — CHANGES_REQUESTED** [review]

File: api/management/api_test.go [hunk 11]
```diff
@@ -250,6 +250,55 @@ func TestConvertUpdatableToKCUser(t *testing.T) {
 	assert.Equal(t, businessID, *ConvertUpdatableToKCUser(user).Attributes.GetString(constants.AttrbBusinessID))
 }
 
+func TestConvertAPICRole(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (basbeu) [reply api/management/api_test.go:253]: Yes this is a typo ! Corrected

File: pkg/management/component.go (hunk 7 above)
Comment (basbeu) [reply pkg/management/component.go:1451]: Very good catch ! Corrected !

File: pkg/management/component.go (hunk 9 above)
Comment (basbeu) [reply pkg/management/component.go:1422]: Corrected

File: api/management/api.go (hunk 8 above)
Comment (basbeu) [reply api/management/api.go:479]: Corrected

File: go.mod (hunk 10 above)
Comment (basbeu) [reply go.mod]: Corrected !

**Review by harture — APPROVED** [review]

**Pull request merged**